### PR TITLE
fix(error): classify DAPI transport errors and decode gRPC domain errors

### PR DESCRIPTION
**Pull request by lklimek**
## Summary

- Replaces the dynamic `sdk_error_user_message()` dispatch function with **17 dedicated `TaskError` variants** for DAPI transport and SDK-level errors, enabling downstream code to pattern-match on error kind
- Decodes gRPC domain errors carried as plain `Internal` status (e.g., Drive's `UniqueKeyAlreadyExists`) via message-based heuristics, with TODOs tracking migration to structured `dash-serialized-consensus-error-bin` decoding
- Distinguishes connection-refused, timeout, and generic unavailability within gRPC `Unavailable` status — each with an accurate, user-facing message
- Auto-dismisses **all** banner types (warning/error after 9s, success/info after 5s) so error messages don't persist indefinitely

### Error classification (`src/backend_task/error.rs`)

- **New `TaskError` variants**: `DapiUnavailable`, `DapiTimeout`, `DapiConnectionRefused`, `DapiInternalError`, `DapiDeadlineExceeded`, `DapiAccessDenied`, `DapiResourceExhausted`, `DapiNoAddresses`, `DapiAllAddressesExhausted`, `SdkTimeout`, `DapiStaleNode`, `PlatformRejected`, `PlatformAlreadyExists`, `OperationCancelled`, `IdentityNonceOverflow`, `IdentityNonceNotFound`
- **Removed** `sdk_error_user_message()` helper — user-facing messages now live in `#[error("...")]` attributes on each variant
- **gRPC `Internal` heuristic**: matches `"a unique key with that hash already exists"`, `"duplicated public key"`, and `"identity key already exists"` to route to `DuplicateIdentityPublicKey` before the consensus-error classifier runs
- **gRPC `Unavailable` sub-classification**: timeout keywords → `DapiTimeout`, connect error/connection refused → `DapiConnectionRefused`, everything else → `DapiUnavailable`
- **Improved doc comments** on `DuplicateIdentityPublicKey` (clarifies uniqueness constraint) and `DuplicateIdentityPublicKeyId` (clarifies per-identity scope)
- Security/TODO comment noting message-based heuristic is spoofable but has no auth/data impact

### Banner auto-dismiss (`src/ui/components/message_banner.rs`)

- Split `DEFAULT_AUTO_DISMISS` into `DEFAULT_AUTO_DISMISS_SHORT` (5s) and `DEFAULT_AUTO_DISMISS_LONG` (9s)
- `Warning` and `Error` banners now auto-dismiss after 9s (previously persisted indefinitely)
- `Success` and `Info` banners remain at 5s

### Tests

- **5 new classification tests**: timeout, unique-key duplicate, generic internal, deadline exceeded, no available addresses
- **7 QA edge-case tests**: unavailable without timeout keywords, resource exhausted, unauthenticated, `NoAvailableAddressesToRetry` at `DapiClientError` level, internal "already exists" without raw gRPC text leaking, internal "duplicate" without identity-key keywords stays `DapiInternalError`, connection-refused via "connect error" substring
- **Updated banner test**: error banner now expects auto-dismiss countdown to be present

## Test plan

- [ ] Run `cargo test` — all 12 new error classification tests should pass
- [ ] Run `cargo test --test kittest` — banner test should pass with updated assertion
- [ ] Manual: trigger a DAPI timeout (e.g., disconnect network) and verify the banner shows "timed out" message and auto-dismisses after 9s
- [ ] Manual: trigger a connection-refused error and verify "Could not reach" message (not "timed out")
- [ ] Manual: verify success banners still dismiss after 5s

<sub>🤖 Co-authored by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent</sub>

### CODE REVIEW

**Comment by lklimek** [comment]
## PR Summary

### Delivered ✅

1. **Specific user messages for DAPI transport errors** — 6 distinct gRPC status code branches instead of one blanket "Could not connect" message (timeout, connection-refused, unavailable, deadline-exceeded, resource-exhausted, auth denial)

2. **Typed classification of gRPC Internal domain errors** — "unique key already exists" errors that arrive as `DapiClientError` are now recognized as `DuplicateIdentityPublicKey` via message-text matching

3. **27 tests** (was 15) — comprehensive coverage of all new paths including edge cases

4. **Security-reviewed** — no information disclosure, spoofability documented with TODO

### Left undone ⚠️

1. **No base64/CBOR metadata decoding** — the `drive-error-data-bin` binary payload in gRPC metadata is still ignored. We match on `status.message()` text as a workaround. This is fragile (message text could change) and won't catch errors where the message is generic but the metadata carries the real cause.
   - **Root cause**: SDK looks for `dash-serialized-consensus-error-bin` metadata key, but Drive sends `drive-error-data-bin` — key mismatch means structured errors are never decoded
   - **No test vectors captured** — need real DAPI responses to build proper tests

2. **Only "unique key already exists" is classified** — other domain errors arriving as gRPC `Internal` (e.g. duplicate documents, credit transfer failures) still fall through to the generic "internal error" message. More patterns can be added incrementally.

3. **IP addresses in Debug chain** (pre-existing) — `tonic::Status` Debug output can include DAPI node IPs in the details panel. Out of scope for this PR.

### Bottom line

The PR fixes the **user-visible symptoms** — no more misleading "Could not connect" for timeouts and duplicate keys. The **root cause** (SDK not decoding `drive-error-data-bin`) is an upstream platform issue.

<sub>🤖 Co-authored by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent</sub>

File: src/backend_task/error.rs [hunk 1]
```diff
@@ -726,6 +789,29 @@ impl From<dashcore_rpc::Error> for TaskError {
...
+        // TODO: replace with structured `dash-serialized-consensus-error-bin` decoding.
+        if let SdkError::DapiClientError(DapiClientError::Transport(TransportError::Grpc(status))) =
+            &error
+            && status.code() == Code::Internal
+        {
+            let msg = status.message().to_lowercase();
+            if msg.contains("unique key") && msg.contains("already exists") {
+                return TaskError::DuplicateIdentityPublicKey {
+                    source_error: Box::new(error),
+                };
```
Comment (thepastaclaw) [inline src/backend_task/error.rs]: **🟡 Suggestion: "unique key with that hash already exists" may map to wrong variant**

The message "a unique key with that hash already exists" originates from GroveDB's unique-index constraint on the identity key hash — this corresponds to `DuplicatedIdentityPublicKeyIdStateError` (key hash collision), not `DuplicatedIdentityPublicKeyStateError` (key data collision). The code maps it to `DuplicateIdentityPublicKey` instead of `DuplicateIdentityPublicKeyId`.

Downgraded from blocking: the user-facing messages are nearly identical ("This public key is already registered" vs "This key is already registered"), so user impact is negligible. The structured consensus error path at lines 833-838 handles the distinction correctly when the SDK successfully decodes the error. This heuristic fallback is inherently imprecise, but if precision is desired, the variant should match the storage-layer semantics.

<details>
<summary>💡 Suggested change</summary>

```suggestion
            if msg.contains("unique key") && msg.contains("already exists") {
                return TaskError::DuplicateIdentityPublicKeyId {
                    source_error: Box::new(error),
                };
            }
```

</details>

<sub>source: ['codex-general', 'codex-rust-quality']</sub>

<details>
<summary>🤖 Fix this with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/backend_task/error.rs`:
- [SUGGESTION] lines 803-806: "unique key with that hash already exists" may map to wrong variant
  The message "a unique key with that hash already exists" originates from GroveDB's unique-index constraint on the identity key hash — this corresponds to `DuplicatedIdentityPublicKeyIdStateError` (key hash collision), not `DuplicatedIdentityPublicKeyStateError` (key data collision). The code maps it to `DuplicateIdentityPublicKey` instead of `DuplicateIdentityPublicKeyId`.

Downgraded from blocking: the user-facing messages are nearly identical ("This public key is already registered" vs "This key is already registered"), so user impact is negligible. The structured consensus error path at lines 833-838 handles the distinction correctly when the SDK successfully decodes the error. This heuristic fallback is inherently imprecise, but if precision is desired, the variant should match the storage-layer semantics.
```
</details>

**Review by thepastaclaw — COMMENTED** [review]
## Code Review

Well-structured PR that adds message-based heuristic classification of gRPC Internal domain errors and differentiates DAPI transport errors into specific user-facing messages. Test coverage is solid (12+ tests). No true blocking issues found — the two findings flagged as blocking by codex agents are either semantically negligible (variant mapping) or upstream SDK limitations (retry wrapping). Several actionable suggestions around message accuracy and test coverage.

*Reviewed commit: 95e6b548*

🟡 5 suggestion(s) | 💬 2 nitpick(s)

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/backend_task/error.rs`:
- [SUGGESTION] lines 803-806: "unique key with that hash already exists" may map to wrong variant
  The message "a unique key with that hash already exists" originates from GroveDB's unique-index constraint on the identity key hash — this corresponds to `DuplicatedIdentityPublicKeyIdStateError` (key hash collision), not `DuplicatedIdentityPublicKeyStateError` (key data collision). The code maps it to `DuplicateIdentityPublicKey` instead of `DuplicateIdentityPublicKeyId`.

Downgraded from blocking: the user-facing messages are nearly identical ("This public key is already registered" vs "This key is already registered"), so user impact is negligible. The structured consensus error path at lines 833-838 handles the distinction correctly when the SDK successfully decodes the error. This heuristic fallback is inherently imprecise, but if precision is desired, the variant should match the storage-layer semantics.
- [SUGGESTION] lines 685-686: Auth-denial message promises automatic server failover that likely won't happen
  For `Unauthenticated | PermissionDenied`, the message says "the app will try a different server." In the DAPI client SDK, `tonic::Status::can_retry()` returns true only for Cancelled, Unknown, DeadlineExceeded, ResourceExhausted, Aborted, Internal, and Unavailable — auth-denial codes are non-retryable. The user will expect automatic failover but the SDK won't perform it. The message should not promise behavior the SDK doesn't provide.
- [SUGGESTION] lines 808-812: Second gRPC Internal classifier branch ("duplicate" + "identity" + "key") has no positive test
  The first branch ("unique key" + "already exists") is tested by `dapi_grpc_internal_unique_key_classifies_as_duplicate`. The negative case is tested by `qa_dapi_grpc_internal_duplicate_without_identity_key_stays_sdk_error`. But no test exercises the second branch with a message matching all three keywords ("duplicate", "identity", "key"), so there's no verification it classifies correctly.
- [SUGGESTION] lines 798-813: SDK retry wrapper can mask domain errors from the gRPC Internal classifier
  This classification only runs for `SdkError::DapiClientError(Transport(Grpc(_)))`. When a gRPC Internal status carrying a domain error (e.g., duplicate key) is retried by the SDK and all retries fail, the last error gets wrapped as `SdkError::NoAvailableAddressesToRetry(Box<Error>)`. In that case, the domain fallback at lines 798-812 is skipped entirely, and the user sees "All servers unreachable" instead of the domain-specific message. This is fundamentally an SDK-level limitation that this PR can't fully solve, but a comment documenting this known gap would help future maintainers. The structured consensus path (lines 822-864) has the same limitation.

In `src/ui/components/message_banner.rs`:
- [SUGGESTION] lines 510-514: Error/warning banners auto-dismissing at 9s may cause users to miss critical errors
  Previously, Error and Warning banners persisted until manual dismissal. This change auto-dismisses them after 9 seconds. For transient issues this is fine, but for errors requiring deliberate user action (wallet locked, insufficient funds, key conflicts), 9 seconds may not be enough time to read, understand, and act. The infrastructure already supports per-banner overrides via `with_auto_dismiss()`, but no callsite currently uses it to extend the timeout for high-importance errors. Consider either: (a) keeping Error banners persistent by default, or (b) documenting a pattern for critical errors to override with a longer/no timeout.
```
</details>

File: src/backend_task/error.rs [hunk 2]
```diff
@@ -647,6 +650,69 @@ fn spv_user_message(e: &crate::spv::SpvError) -> &'static str {
...
+                "The Dash network returned an internal error. Please retry in a few moments."
+                    .to_string()
+            }
+        }
+        Code::DeadlineExceeded => {
+            "The operation took too long. Please retry — it often succeeds on the next attempt."
+                .to_string()
+        }
+        Code::Unauthenticated | Code::PermissionDenied => {
+            "Access was denied by the network server. Please retry — the app will try a different server.".to_string()
```
Comment (thepastaclaw) [inline src/backend_task/error.rs]: **🟡 Suggestion: Auth-denial message promises automatic server failover that likely won't happen**

For `Unauthenticated | PermissionDenied`, the message says "the app will try a different server." In the DAPI client SDK, `tonic::Status::can_retry()` returns true only for Cancelled, Unknown, DeadlineExceeded, ResourceExhausted, Aborted, Internal, and Unavailable — auth-denial codes are non-retryable. The user will expect automatic failover but the SDK won't perform it. The message should not promise behavior the SDK doesn't provide.

<details>
<summary>💡 Suggested change</summary>

```suggestion
        Code::Unauthenticated | Code::PermissionDenied => {
            "Access was denied by the network server. Please retry in a few moments.".to_string()
        }
```

</details>

<sub>source: ['codex-general']</sub>

<details>
<summary>🤖 Fix this with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/backend_task/error.rs`:
- [SUGGESTION] lines 685-686: Auth-denial message promises automatic server failover that likely won't happen
  For `Unauthenticated | PermissionDenied`, the message says "the app will try a different server." In the DAPI client SDK, `tonic::Status::can_retry()` returns true only for Cancelled, Unknown, DeadlineExceeded, ResourceExhausted, Aborted, Internal, and Unavailable — auth-denial codes are non-retryable. The user will expect automatic failover but the SDK won't perform it. The message should not promise behavior the SDK doesn't provide.
```
</details>

File: src/ui/components/message_banner.rs [hunk 3]
```diff
@@ -505,11 +506,11 @@ impl Component for MessageBanner {
...
 /// Returns the default auto-dismiss duration for a message type.
-/// `Success` and `Info` auto-dismiss; `Error` and `Warning` persist.
+/// All banners auto-dismiss: success/info after 5s, warning/error after 9s.
 fn default_auto_dismiss(message_type: MessageType) -> Option<Duration> {
     match message_type {
-        MessageType::Success | MessageType::Info => Some(DEFAULT_AUTO_DISMISS),
-        MessageType::Error | MessageType::Warning => None,
+        MessageType::Success | MessageType::Info => Some(DEFAULT_AUTO_DISMISS_SHORT),
+        MessageType::Warning | MessageType::Error => Some(DEFAULT_AUTO_DISMISS_LONG),
     }
```
Comment (thepastaclaw) [inline src/ui/components/message_banner.rs:514]: **🟡 Suggestion: Error/warning banners auto-dismissing at 9s may cause users to miss critical errors**

Previously, Error and Warning banners persisted until manual dismissal. This change auto-dismisses them after 9 seconds. For transient issues this is fine, but for errors requiring deliberate user action (wallet locked, insufficient funds, key conflicts), 9 seconds may not be enough time to read, understand, and act. The infrastructure already supports per-banner overrides via `with_auto_dismiss()`, but no callsite currently uses it to extend the timeout for high-importance errors. Consider either: (a) keeping Error banners persistent by default, or (b) documenting a pattern for critical errors to override with a longer/no timeout.

<sub>source: ['claude-general', 'claude-rust-quality', 'codex-rust-quality']</sub>

<details>
<summary>🤖 Fix this with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/ui/components/message_banner.rs`:
- [SUGGESTION] lines 510-514: Error/warning banners auto-dismissing at 9s may cause users to miss critical errors
  Previously, Error and Warning banners persisted until manual dismissal. This change auto-dismisses them after 9 seconds. For transient issues this is fine, but for errors requiring deliberate user action (wallet locked, insufficient funds, key conflicts), 9 seconds may not be enough time to read, understand, and act. The infrastructure already supports per-banner overrides via `with_auto_dismiss()`, but no callsite currently uses it to extend the timeout for high-importance errors. Consider either: (a) keeping Error banners persistent by default, or (b) documenting a pattern for critical errors to override with a longer/no timeout.
```
</details>

File: src/backend_task/error.rs [hunk 4]
```diff
@@ -726,6 +789,29 @@ impl From<dashcore_rpc::Error> for TaskError {
...
+            if msg.contains("unique key") && msg.contains("already exists") {
+                return TaskError::DuplicateIdentityPublicKey {
+                    source_error: Box::new(error),
+                };
+            }
+            if msg.contains("duplicate") && msg.contains("identity") && msg.contains("key") {
+                return TaskError::DuplicateIdentityPublicKey {
+                    source_error: Box::new(error),
+                };
+            }
```
Comment (thepastaclaw) [inline src/backend_task/error.rs]: **🟡 Suggestion: Second gRPC Internal classifier branch ("duplicate" + "identity" + "key") has no positive test**

The first branch ("unique key" + "already exists") is tested by `dapi_grpc_internal_unique_key_classifies_as_duplicate`. The negative case is tested by `qa_dapi_grpc_internal_duplicate_without_identity_key_stays_sdk_error`. But no test exercises the second branch with a message matching all three keywords ("duplicate", "identity", "key"), so there's no verification it classifies correctly.

<details>
<summary>💡 Suggested change</summary>

```suggestion
#[test]
fn dapi_grpc_internal_duplicate_identity_key_classifies_as_duplicate() {
    let status = dash_sdk::dapi_grpc::tonic::Status::internal(
        "duplicate identity public key found",
    );
    let dapi_err = DapiClientError::Transport(TransportError::Grpc(status));
    let sdk_err = SdkError::DapiClientError(dapi_err);
    let err = TaskError::from(sdk_err);
    assert!(
        matches!(err, TaskError::DuplicateIdentityPublicKey { .. }),
        "Expected DuplicateIdentityPublicKey for duplicate+identity+key message, got: {err:?}"
    );
}
```

</details>

<sub>source: ['claude-rust-quality']</sub>

<details>
<summary>🤖 Fix this with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/backend_task/error.rs`:
- [SUGGESTION] lines 808-812: Second gRPC Internal classifier branch ("duplicate" + "identity" + "key") has no positive test
  The first branch ("unique key" + "already exists") is tested by `dapi_grpc_internal_unique_key_classifies_as_duplicate`. The negative case is tested by `qa_dapi_grpc_internal_duplicate_without_identity_key_stays_sdk_error`. But no test exercises the second branch with a message matching all three keywords ("duplicate", "identity", "key"), so there's no verification it classifies correctly.
```
</details>

File: src/backend_task/error.rs [hunk 5]
```diff
@@ -726,6 +789,29 @@ impl From<dashcore_rpc::Error> for TaskError {
...
+        }
```
Comment (thepastaclaw) [inline src/backend_task/error.rs:848]: **🟡 Suggestion: SDK retry wrapper can mask domain errors from the gRPC Internal classifier**

This classification only runs for `SdkError::DapiClientError(Transport(Grpc(_)))`. When a gRPC Internal status carrying a domain error (e.g., duplicate key) is retried by the SDK and all retries fail, the last error gets wrapped as `SdkError::NoAvailableAddressesToRetry(Box<Error>)`. In that case, the domain fallback at lines 798-812 is skipped entirely, and the user sees "All servers unreachable" instead of the domain-specific message. This is fundamentally an SDK-level limitation that this PR can't fully solve, but a comment documenting this known gap would help future maintainers. The structured consensus path (lines 822-864) has the same limitation.

<sub>source: ['codex-rust-quality']</sub>

<details>
<summary>🤖 Fix this with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/backend_task/error.rs`:
- [SUGGESTION] lines 798-813: SDK retry wrapper can mask domain errors from the gRPC Internal classifier
  This classification only runs for `SdkError::DapiClientError(Transport(Grpc(_)))`. When a gRPC Internal status carrying a domain error (e.g., duplicate key) is retried by the SDK and all retries fail, the last error gets wrapped as `SdkError::NoAvailableAddressesToRetry(Box<Error>)`. In that case, the domain fallback at lines 798-812 is skipped entirely, and the user sees "All servers unreachable" instead of the domain-specific message. This is fundamentally an SDK-level limitation that this PR can't fully solve, but a comment documenting this known gap would help future maintainers. The structured consensus path (lines 822-864) has the same limitation.
```
</details>

File: src/backend_task/error.rs [hunk 6]
```diff
@@ -647,6 +650,69 @@ fn spv_user_message(e: &crate::spv::SpvError) -> &'static str {
...
+            "The operation took too long. Please retry — it often succeeds on the next attempt."
+                .to_string()
+        }
+        Code::Unauthenticated | Code::PermissionDenied => {
+            "Access was denied by the network server. Please retry — the app will try a different server.".to_string()
+        }
+        Code::ResourceExhausted => {
+            "The network server is overloaded. Please wait a moment and retry.".to_string()
+        }
+        _ => "Could not connect to the Dash network. Please retry in a few moments.".to_string(),
```
Comment (thepastaclaw) [inline src/backend_task/error.rs]: **💬 Nitpick: Wildcard arm says "Could not connect" for all unhandled gRPC codes**

The `_ =>` arm produces "Could not connect to the Dash network" for any unmatched gRPC status code, including codes like NotFound or InvalidArgument that aren't connection failures. While these codes are unlikely in the DAPI context, a more neutral wording would be more accurate.

<details>
<summary>💡 Suggested change</summary>

```suggestion
        _ => "A network error occurred. Please retry in a few moments.".to_string(),
```

</details>

<sub>source: ['claude-general']</sub>

File: src/backend_task/error.rs [hunk 7]
```diff
@@ -647,6 +650,69 @@ fn spv_user_message(e: &crate::spv::SpvError) -> &'static str {
...
+        DapiClientError::Transport(TransportError::Grpc(status)) => {
+            grpc_status_user_message(status)
+        }
+        DapiClientError::NoAvailableAddresses => {
+            "No Dash network servers are configured. Please check your network settings."
+                .to_string()
+        }
+        DapiClientError::NoAvailableAddressesToRetry(_) => {
+            "All Dash network servers are temporarily unreachable. Please wait a minute and retry."
+                .to_string()
```
Comment (thepastaclaw) [inline src/backend_task/error.rs]: **💬 Nitpick: Duplicated message string for NoAvailableAddressesToRetry**

The exact same message "All Dash network servers are temporarily unreachable. Please wait a minute and retry." appears in both `dapi_client_error_user_message` (line 708) and `sdk_error_user_message` (line 743). Consider extracting to a constant to keep wording in sync.

<sub>source: ['claude-general']</sub>

File: src/backend_task/error.rs
Comment (lklimek) [inline src/backend_task/error.rs]: Note this key must be uniquer; platform also supports non-unique keys.

Comment (lklimek) [inline src/backend_task/error.rs]: Clarify this is about key ID for this particular identity.

File: src/backend_task/error.rs [hunk 8]
```diff
@@ -745,6 +808,29 @@ impl From<dashcore_rpc::Error> for TaskError {
...
+            &error
+            && status.code() == Code::Internal
+        {
+            let msg = status.message().to_lowercase();
+            if msg.contains("unique key") && msg.contains("already exists") {
+                return TaskError::DuplicateIdentityPublicKey {
+                    source_error: Box::new(error),
+                };
+            }
+            if msg.contains("duplicate") && msg.contains("identity") && msg.contains("key") {
```
Comment (lklimek) [inline src/backend_task/error.rs]: this match is too general and risks false positives; do more explicit match if possible, find and add a comment with  original error msg, as generated by drive; note this is a workaround and mark TODO.

File: src/backend_task/error.rs [hunk 9]
```diff
@@ -745,6 +808,29 @@ impl From<dashcore_rpc::Error> for TaskError {
...
+        // metadata, but some platform errors arrive as plain Internal with descriptive
+        // message text only. This is a message-based workaround (see also #714 fallback).
+        // NOTE: a malicious node could spoof these messages; no auth/data impact.
+        // TODO: replace with structured `dash-serialized-consensus-error-bin` decoding.
+        if let SdkError::DapiClientError(DapiClientError::Transport(TransportError::Grpc(status))) =
+            &error
+            && status.code() == Code::Internal
+        {
+            let msg = status.message().to_lowercase();
+            if msg.contains("unique key") && msg.contains("already exists") {
```
Comment (lklimek) [inline src/backend_task/error.rs]: this match is too general and risks false positives; do more explicit match if possible, find and add a comment with  original error msg, as generated by drive; note this is a workaround and mark TODO.

File: src/backend_task/error.rs [hunk 10]
```diff
@@ -178,10 +185,133 @@ pub enum TaskError {
     #[error("Internal update failed. Please retry the operation.")]
     InternalSendError,
 
+    /// DAPI server is temporarily unavailable (gRPC Unavailable).
+    #[error(
+        "A Dash network server is temporarily unavailable. Please retry — the app will try a different server."
```
Comment (lklimek) [inline src/backend_task/error.rs]:
```suggestion
        "A Dash network server is temporarily unavailable. Please retry."
```

File: src/backend_task/error.rs [hunk 11]
```diff
@@ -178,10 +185,133 @@ pub enum TaskError {
...
+    )]
+    DapiUnavailable {
+        #[source]
+        source_error: Box<SdkError>,
+    },
+
+    /// Connection to DAPI server timed out (gRPC Unavailable with timeout message).
+    #[error(
+        "Connection to a Dash network server timed out. Please retry — the app will try a different server."
```
Comment (lklimek) [inline src/backend_task/error.rs]:
```suggestion
        "Connection to a Dash network server timed out. Please retry."
```

File: src/backend_task/error.rs [hunk 12]
```diff
@@ -178,10 +185,133 @@ pub enum TaskError {
...
+    )]
+    DapiTimeout {
+        #[source]
+        source_error: Box<SdkError>,
+    },
+
+    /// Could not reach DAPI server (gRPC Unavailable with connection refused).
+    #[error(
+        "Could not reach a Dash network server. Please retry — the app will try a different server."
```
Comment (lklimek) [inline src/backend_task/error.rs]:
```suggestion
        "Could not reach a Dash network server. Please retry."
```

File: src/backend_task/error.rs [hunk 13]
```diff
@@ -178,10 +185,133 @@ pub enum TaskError {
...
+
+    /// DAPI deadline exceeded (gRPC DeadlineExceeded).
+    #[error("The operation took too long. Please retry — it often succeeds on the next attempt.")]
+    DapiDeadlineExceeded {
+        #[source]
+        source_error: Box<SdkError>,
+    },
+
+    /// Access denied by DAPI server (gRPC Unauthenticated/PermissionDenied).
+    #[error("Access was denied by the network server. Please retry in a few moments.")]
```
Comment (lklimek) [inline src/backend_task/error.rs]:
```suggestion
    #[error("Access was denied by the network server. Check you password in settings.")]
```

File: src/backend_task/error.rs [hunk 14]
```diff
@@ -178,10 +185,133 @@ pub enum TaskError {
...
+    )]
+    SdkTimeout {
+        timeout_secs: u64,
+        #[source]
+        source_error: Box<SdkError>,
+    },
+
+    /// Connected server is behind (SdkError::StaleNode).
+    #[error(
+        "The server you connected to is behind. Please retry — the app will pick a different server automatically."
```
Comment (lklimek) [inline src/backend_task/error.rs]:
```suggestion
        "The server you connected to is behind. Please retry."
```

File: src/backend_task/error.rs [hunk 15]
```diff
@@ -1044,6 +1249,93 @@ mod tests {
...
+            "Expected DapiTimeout, got: {err:?}"
+        );
+        let msg = err.to_string();
+        assert!(
+            msg.contains("timed out"),
+            "Expected timeout message, got: {msg}"
+        );
+        assert!(
+            msg.contains("different server"),
+            "Expected retry hint, got: {msg}"
```
Comment (thepastaclaw) [inline src/backend_task/error.rs:1263]: **🔴 Blocking: Test fails: stale `contains("different server")` assertion on DapiTimeout message**

The `DapiTimeout` message was shortened from "…Please retry — the app will try a different server" to "…Please retry." in commit 87c3576, but this test still asserts `msg.contains("different server")`. Confirmed failing: `cargo test --lib -p dash-evo-tool -- error::tests` panics here with "Expected retry hint, got: Connection to a Dash network server timed out. Please retry."

<details>
<summary>💡 Suggested change</summary>

```suggestion
        assert!(
            msg.contains("retry"),
            "Expected retry hint, got: {msg}"
        );
```

</details>

<sub>source: ['claude', 'codex']</sub>

<details>
<summary>🤖 Fix this with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/backend_task/error.rs`:
- [BLOCKING] lines 1268-1271: Test fails: stale `contains("different server")` assertion on DapiTimeout message
  The `DapiTimeout` message was shortened from "…Please retry — the app will try a different server" to "…Please retry." in commit 87c3576, but this test still asserts `msg.contains("different server")`. Confirmed failing: `cargo test --lib -p dash-evo-tool -- error::tests` panics here with "Expected retry hint, got: Connection to a Dash network server timed out. Please retry."
```
</details>

File: src/backend_task/error.rs [hunk 16]
```diff
@@ -1053,4 +1345,154 @@ mod tests {
...
+            "Expected DapiUnavailable, got: {err:?}"
+        );
+        let msg = err.to_string();
+        assert!(
+            msg.contains("temporarily unavailable"),
+            "Expected unavailable message without timeout text, got: {msg}"
+        );
+        assert!(
+            msg.contains("different server"),
+            "Expected retry hint, got: {msg}"
```
Comment (thepastaclaw) [inline src/backend_task/error.rs:1361]: **🔴 Blocking: Test fails: stale `contains("different server")` assertion on DapiUnavailable message**

Same issue as the DapiTimeout test. The `DapiUnavailable` message is now "A Dash network server is temporarily unavailable. Please retry." — no "different server" text. Confirmed failing with the same `cargo test` run.

<details>
<summary>💡 Suggested change</summary>

```suggestion
        assert!(
            msg.contains("retry"),
            "Expected retry hint, got: {msg}"
        );
```

</details>

<sub>source: ['claude', 'codex']</sub>

<details>
<summary>🤖 Fix this with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/backend_task/error.rs`:
- [BLOCKING] lines 1366-1369: Test fails: stale `contains("different server")` assertion on DapiUnavailable message
  Same issue as the DapiTimeout test. The `DapiUnavailable` message is now "A Dash network server is temporarily unavailable. Please retry." — no "different server" text. Confirmed failing with the same `cargo test` run.
```
</details>

File: src/backend_task/error.rs [hunk 17]
```diff
@@ -178,10 +185,133 @@ pub enum TaskError {
...
+
+    /// DAPI deadline exceeded (gRPC DeadlineExceeded).
+    #[error("The operation took too long. Please retry — it often succeeds on the next attempt.")]
+    DapiDeadlineExceeded {
+        #[source]
+        source_error: Box<SdkError>,
+    },
+
+    /// Access denied by DAPI server (gRPC Unauthenticated/PermissionDenied).
+    #[error("Access was denied by the network server. Check you password in settings.")]
```
Comment (thepastaclaw) [inline src/backend_task/error.rs]: **🔴 Blocking: Typo in DapiAccessDenied user-facing message: "you password" → "your password"**

The `#[error]` attribute reads "Check you password in settings." — missing the 'r' in "your". This is a user-facing message shown in the MessageBanner.

<details>
<summary>💡 Suggested change</summary>

```suggestion
    #[error("Access was denied by the network server. Check your password in settings.")]
```

</details>

<sub>source: ['claude', 'codex']</sub>

<details>
<summary>🤖 Fix this with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/backend_task/error.rs`:
- [BLOCKING] line 230: Typo in DapiAccessDenied user-facing message: "you password" → "your password"
  The `#[error]` attribute reads "Check you password in settings." — missing the 'r' in "your". This is a user-facing message shown in the MessageBanner.
```
</details>

**Review by thepastaclaw — CHANGES_REQUESTED** [review]
## Code Review

The refactoring from runtime string-dispatch functions to dedicated TaskError variants is architecturally sound — every SDK error now gets a typed variant with compile-time `#[error]` messages instead of runtime `sdk_error_user_message()` lookups. The From<SdkError> impl is well-structured with consensus errors checked first, then gRPC transport, then SDK-level variants. However, two test assertions now fail because error messages were shortened in commit 87c35760 ('Apply suggestions from code review') without updating the corresponding `contains("different server")` assertions. There is also a typo in the DapiAccessDenied user-facing message.

*Reviewed commit: cdbfea39*

🔴 3 blocking

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/backend_task/error.rs`:
- [BLOCKING] lines 1268-1271: Test fails: stale `contains("different server")` assertion on DapiTimeout message
  The `DapiTimeout` message was shortened from "…Please retry — the app will try a different server" to "…Please retry." in commit 87c3576, but this test still asserts `msg.contains("different server")`. Confirmed failing: `cargo test --lib -p dash-evo-tool -- error::tests` panics here with "Expected retry hint, got: Connection to a Dash network server timed out. Please retry."
- [BLOCKING] lines 1366-1369: Test fails: stale `contains("different server")` assertion on DapiUnavailable message
  Same issue as the DapiTimeout test. The `DapiUnavailable` message is now "A Dash network server is temporarily unavailable. Please retry." — no "different server" text. Confirmed failing with the same `cargo test` run.
- [BLOCKING] line 230: Typo in DapiAccessDenied user-facing message: "you password" → "your password"
  The `#[error]` attribute reads "Check you password in settings." — missing the 'r' in "your". This is a user-facing message shown in the MessageBanner.
```
</details>

**Review by thepastaclaw — COMMENTED** [review]
## Code Review

Automated code review.

*Reviewed commit: cc4ebb4a*
